### DISCUSSION TITLE
Support private go modules on custom GitLab EE host

### DISCUSSION THREAD

**Discussion by montagao** [renovatebot/renovate#22152] [Suggest an Idea]
**What would you like Renovate to be able to do?**
 We are running Renovate pro on a private GitLab EE instance, which has various go modules and shared dependencies among them. We would like Renovate to be able to fetch dependency updates for these private go modules.

  **Describe the solution you'd like** 
Basically we would like following requirements:
* Renovate should fetch be able to fetch dependency information from go modules that are from an internal repo on a custom GitLab EE host (the same host as the API endpoint)
* Renovate should be able to fetch appropriate update artifacts from go modules that are from an internal repo on a custom GitLab EE host when updating public go modules, by the same token as above.

It is similar to #3498 but for a self-hosted GitLab instance using a Self-Hosted RenovateBot. 

**Describe alternatives you've considered**  
We noticed that we can configure Renovate pro to use a globally installed go binary instead of a docker image to run go, by setting `binarySource` to `global`.   Then we configured Renovate’s  `.gitconfig` to use `RENOVATE_TOKEN`, by running `git config --global url."https://oauth2:${RENOVATE_TOKEN}@${GITLAB_HOST}”.insteadOf “${GITLABHOST}”` in our Renovate container, where `GITLAB_HOST` is supplied by us. 

**Additional context**
We are running Renovate as a k8s deployment.

**Comment by rarkins** [comment]
Can you confirm if Renovate correctly handles the same setup (i.e. module on a separate GitLab host) if *no* credentials are required? I want to make sure that we actually handle GitLab hosts correctly in the first place, before then concentrating on the authentication part.

**Comment by dvdizon** [comment]
Hi @rarkins, we unfortunately do not have an environment / setup with GitLab hosts that require no credentials required.

**Comment by R-omk** [comment]
To fix this issue, it need to remove all unnecessary manipulations with the git tags:
https://github.com/renovatebot/renovate/blob/cbe1246/lib/datasource/go/index.ts#L88-L90

To fetch new dependency for go modules  just need to run  native go  command:
```
go get  -u=patch ./...  # for patch only

go get  -u ./... # also for minor versions
```
its will  look like 
```
go: gitlab.lolcal.loc/root/test_dependency patch => v0.6.2
```

then  just again parse go.mod 

@rarkins 


related:
https://github.com/renovatebot/renovate/pull/2584

**Comment by R-omk** [comment]
^^  Such a solution will be completely independent of the source storage method and is preferred.
`go get`  will do everything himself, and even much better than manually searching for dependencies by tags in a git.

All problems with authorization are easily solved as described in the header of this topic:

.gitconfig
```
[url "https://oauth2:SOME_SECRET_TOKEN_FORM_GITLAB@gitlab.lolcal.loc/"]
        insteadOf = https://gitlab.lolcal.loc/
```

Optionally, for self-signed certificates, you can add a special flag  `go get -insecure   -u ./...`

**Comment by rarkins** [comment]
From my experience implementing dozens of managers, any time we've used binaries for lookups and then attempted to parse the results, it's been very painful. And trust me, the following is never ever true, regardless of manager:

> All problems with authorization are easily solved

I am not encouraged either by the rapid changing of Go's modules implementation, if we are to be the one interpreting it and trying to keep up - ideally we could wait until it's more stable.

However, if you're interested to write a proof of concept PR to demonstrate that this approach works, I would definitely consider it.

**Comment by R-omk** [comment]
Currently, the go modules look better and more stable than a year ago.

`go get` now has more advantages, for example:

https://github.com/golang/go/wiki/Modules#daily-workflow
>Using a branch name such as go get foo@master (foo@tip with mercurial) is one way to obtain the latest commit regardless of whether or not it has a semver tag.

It can build a stable version from a branch name natively supported in the go module.

go.mod become look like:
```
require gitlab.lolcal.loc/root/test_dependency v0.6.3-0.20200305142108-db0c56da61bf
```

This is a very useful feature for the renovatebot that automates the development and testing process.

**You don't need** to _parse the results_  because you already implemented this parser.
`go get  -u=patch ./...`  this command overwrites the existing go.module file, you just need to repeatedly extract the finished result from it

**Comment by rarkins** [comment]
Happy to receive a PR if you have time. Remember that Renovate does a lot more than simple upgrades though. For example if a dependency has a patch upgrade, a minor upgrade, and two major upgrades then it's entirely possible to configure Renovate to receive four separate PRs to choose from (but by default it's maximum two - the latest major and latest non-major).

BTW I had hoped that using the new Go proxy as a kind of API lookup for new versions would have solved most of the problem, although possibly not for local dependencies.

**Comment by R-omk** [comment]
Of course, I understand that he can make several separate PRs:

For example, it can even make three requests for each package individually:

```
go get gitlab.lolcal.loc/root/test_dependency@v0.6  # for  patch
go get gitlab.lolcal.loc/root/test_dependency@v0  # for  minior
go get gitlab.lolcal.loc/root/test_dependency@master  # forlatest commit
```
The last example is good for nightly builds. 

As an optimization, it can simultaneously get all possible versions using two commands (below) and then make separate PR for each package if necessary
```
go get  -u=patch ./...  # for patch only
go get  -u ./... # also for minor versions
```
For local repositories, we explicitly disable proxies like this

~/.bashrc
```
export GOPRIVATE=*.lolcal.loc`
```

**Comment by HonkingGoose** [comment]
@rarkins said earlier:
> I am not encouraged either by the rapid changing of Go's modules implementation, if we are to be the one interpreting it and trying to keep up - ideally we could wait until it's more stable.
> 
> However, if you're interested to write a proof of concept PR to demonstrate that this approach works, I would definitely consider it.

@rarkins What do you want to do with this issue? This issue has not seen any real progress since march 2020, so I'm wondering if this is still a thing you want to work on or not?

I've marked it `priority-4-low` for now, as it's not seen any movement recently.

**Comment by malmor** [comment]
Hey there, we are facing this "feature gap" aswell.

We are running a self-hosted Gitlab CE instance with some go modules inside.
Renovate is running inside Gitlab CI/CD pipelines - this is working really great 👍

But after starting with our go projects we realised that our private modules can not be processed.

Info log says:
`Lock file error - aborting (repository=my-group/my-project)`

The debug log shows the login problem:
`"Command failed: docker run --rm --name=renovate_go --label=renovate_child ... docker.io/renovate/go:1.16.3 bash -l -c \"git config --global url.\\\"https://**redacted**@github.com/\\\".insteadOf \\\"https://**redacted**@v0.0.0-20210214151740-6d496a9199a6: invalid version: git fetch -f origin refs/heads/*:refs/heads/* refs/tags/*:refs/tags/* in /builds/renovate/renovate-runner/renovate/cache/others/go/pkg/mod/cache/vcs/e36a81276a43d170c6b8250d67389c54d44940b236ec159329b940f77074cc64: exit status 128:\n\tfatal: could not read Username for 'https://git.enterprise.local': terminal prompts disabled\n",`

We tried setting our own .gitconfig entry (`git config --global url."https://gitlab-ci-token:$CI_JOB_TOKEN@git.enterprise-local/".insteadOf "https://git.enterprise-local/"`) inside the renovate dind container, but this settings isn't forwarded towards the go service container.


Looking forward to hearing some updates on this 👍

**Comment by rarkins** [comment]
The conversation for this feature request went a bit off the rails, and hasn't progressed past requirements, so I have converted it to an Ideas Discussion. If someone can get this back on track then we can create a new, clean feature request issue.
